### ISSUE TITLE
decorate feature service attachement responses w/ full urls?

### ISSUE THREAD

**Issue by tomwayson** [Esri/arcgis-rest-js#371]
@jPurush was going to add this code that we use in the Hub to Hub.js:

https://github.com/Esri/hub.js/blob/2ada2dfe36253e3fa5b75d4993ff28c41198d45f/packages/events/src/get.ts#L45-L60

I wonder if instead of adding it there if rest-js should be decorating  the [getAttachments()](https://esri.github.io/arcgis-rest-js/api/feature-service/getAttachments/) responses w/ the full url similar to the the suggestion for thumbnail urls #323?

As w/ that issue, I'm a bit wary of appending the token. Also it looks like the `?v=` parameter is not necessarily supported by [the API](https://developers.arcgis.com/rest/services-reference/attachment-feature-service-.htm), but is something we added in the Hub for cache busting. Not sure if we should add that in all cases.

Alternatively, we could provide a `getAttachmentUrl(featureServiceUrl: string, attachmentId: number, authentication?: IAuthenticationManager)` util fn and let people `.map()` over the attachment infos to get the URLs themselves?

**Comment by jgravois** [comment]
> I wonder if instead of adding it there if rest-js should be decorating `getAttachments()` responses w/ the full url 

yes. i am _all for_ this.

i'm also willing to do the testing to quell your concerns about appending tokens because i think it's an appropriate default.

**Comment by itrew** [comment]
Saw this issue and thought I'd chime in if that's alright. As with @tomwayson I'd be wary about appending the token, as it encourages users to be a little more lax with security. For example, we had a node express app that was developed to display the groups users belonged to. The app rendered a page using Pug and included the group's thumbnail url with a token appended. Unfortunately the token was the same one being used on the back-end for other organizational requests (e.g. admin token). So someone only needed to access the page source to find a current admin token.

In other words, if you have a server app using this lib making requests with an account with elevated privileges, it makes it uncomfortably easy to let an admin token slip through if it's appended to an arbitrary URL.

**Comment by tomwayson** [comment]
Thanks for the feedback @itrew 

We could always add a `outputUrl: boolean | string` option to [IGetAttachmentsOptions](https://esri.github.io/arcgis-rest-js/api/feature-service/IGetAttachmentsOptions/). If it's truthy, add a `url` to the attachmentInfos in the response. If `outputUrl === 'token'` then also append the token? Or if that's too cheesy we could always output the URL and just have an `appendToken: boolean` option.

**Comment by itrew** [comment]
I see the value in appending the url by default in almost all instances as that seems to be the most sought after information. So to me it would make sense to just decorate the attachment info with the `url` property, making me lean towards the `appendToken: boolean` option like you've mentioned. I do, however, think that it needs to be **false** by default in order to guard against the situations like in my comment above. That way, it forces devs to consider the security consequences (that should be clearly expressed in documentation) of appending the token to the url.

This is a funny situation, because in a client side application where it's restricted to one user, it makes complete sense to always just append the token for convenience 100% of the time as there is practically no risk. It's just on the server side where it has serious implications, making me think it would be prudent to err on the side of caution for the lib as a whole.

Just my 2 cents. And BTW, thanks guys for putting in the work on this lib. I'm very excited with where it's going 👍.
